README.md
### Switchless Mods for SEGA MasterSystem and MegaDrive
---

This is a collection of earlier projects. These projects are not under development anymore. Documentation is rather simple and can be found in source code comments in most cases.

Please don't write me for any support. Please reach out to community forums and use the knowledge and experiences there.

Best Regards, Peter


MD_BuildInGame/CHANGELOG
CHANGELOG:  
  
Version 1.4:  
 - initial commit

MD_IGR/md_igr_simple.asm
#include  <p12f629.inc>

; -----------------------------------------------------------------------
;   A simple IGR code for the Mega Drive designed to run on a PIC12F629
;
;   Copyright (C) 2016 by Peter Bartmann
;
;   This program is free software; you can redistribute it and/or modify
;   it under the terms of the GNU General Public License as published by
;   the Free Software Foundation; version 2 of the License only.
;
;   This program is distributed in the hope that it will be useful,
;   but WITHOUT ANY WARRANTY; without even the implied warranty of
;   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;   GNU General Public License for more details.
;
;   You should have received a copy of the GNU General Public License
;   along with this program; if not, write to the Free Software
;   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
;
; -----------------------------------------------------------------------
;
;   This program is designed to run on a PIC 12F629 microcontroller.
;
;   pin configuration:
;
;                                 ,----__----.
;                             +5V |1        8| GND
;                   Ext.Mode [in] |2  G5 G0 7| [in/out] Reset Button
;     Ctrl.Port Pin 6 (B+A)  [in] |3  G4 G1 6| [in/out] CPU /HALT
;     Ctrl.Port Pin 9 (C+St) [in] |4  G3 G2 5| [in] n.c. 
;                                 `----------'
;
;  Special Features:
;
;    Ext.Mode (Pin 2)
;      - pin set to low: extended mode is switched off.
;          The microcontroller simply resets the the console for a few
;          milliseconds if the buttons St+A+B+C are held for 1.5seconds at the
;          controller pad.
;      - pin set to high: extended mode is switched on.
;          The microcontroller outputs a reset if the buttons St+A+B+C are held
;          for 1.5s at the controller pad. The reset is then held as long as the
;          button combination is held.
;          This mode can be used if you have a switchless mod installed in your
;          Mega Drive to switch between the modes there.
;          Further, the CPU is paused during that time such that the game does
;          not run if you have pin 3 of the microcontroller connected.
;
;    CPU /HALT (pin 6)
;      Used to pause the game during. The game is only paused in the extended
;      mode (see description above) after St+A+B+C is pressed for 1.5 seconds.
;      CPU /HALT is also triggered if a push on the reset button is triggered.
;      This allows the user, if a switchless mod is installed, to change the
;      mode without the game is running.
;      Connect this pin to:
;        - MD1 CPU (DIL package) pin 17
;        - MD2 CPU (PLCC package) pin 19
;      if this feature is wanted. If this feature is not wanted, leave this pin
;      unconnected.
;
;    Reset Button (pin 7)
;      This pin has to be connected to the reset button. The code will sense how
;      the button is connected, i.e., either as an low-active or high-active
;      reset.
;   
;
;
; Configuration bits: adapt to your setup and needs
    __CONFIG _INTRC_OSC_NOCLKOUT & _WDT_OFF & _PWRTE_ON & _MCLRE_OFF & _BODEN_ON & _CP_OFF & _CPD_OFF

Calibrate_OSCCAL  set 1 ; 0 = no calibration, 1 = with calibration

; -----------------------------------------------------------------------
; macros and definitions

M_movlf macro   literal, toReg  ; move literal to filereg
        movlw   literal
        movwf   toReg
        endm

M_delay_x10ms   macro   literal ; delay about literal x 10ms
                movlw   literal
                movwf   reg_repetition_cnt
                call    delay_x10ms
                endm

M_T1reset   macro   ; reset and start timer1
            clrf    TMR1L
            clrf    TMR1H
            clrf    PIR1
            bsf     T1CON, TMR1ON
            endm

M_push_reset    macro             ; macro pauses CPU, too
                banksel TRISIO
                movlw   0x3c
                movwf   TRISIO    ; in in in in out out
                banksel GPIO
                movfw   reg_reset_type
                movwf   GPIO
                endm

M_release_reset macro             ; macro runs CPU, too
                movfw   reg_reset_type
                xorlw   (1<<RST_BUTTON) ^ (1<<CPU_NHALT)
                movwf   GPIO
                banksel TRISIO
                movlw   0x3f
                movwf   TRISIO    ; in in in in in in
                banksel GPIO
                endm

M_pause_CPU macro
            banksel TRISIO
            bcf     TRISIO, CPU_NHALT
            banksel GPIO
            bcf     GPIO, CPU_NHALT
            endm

M_run_CPU macro
          bsf     GPIO, CPU_NHALT
          banksel TRISIO
          bsf     TRISIO, CPU_NHALT
          banksel GPIO
          endm

M_skipnext_rst_pressed  macro
                        movfw   GPIO
                        xorwf   reg_reset_type, 0
                        andlw   (1<<RST_BUTTON)
                        btfss   STATUS, Z
                        endm

M_skipnext_rst_notpressed   macro
                            movfw   GPIO
                            xorwf   reg_reset_type, 0
                            andlw   (1<<RST_BUTTON)
                            btfsc   STATUS, Z
                            endm

; -----------------------------------------------------------------------
; bits and registers and more

RST_BUTTON  EQU 0
CPU_NHALT   EQU 1
CTRL_P9     EQU 3
CTRL_P6     EQU 4
EXTMODE     EQU 5

reg_reset_type      EQU 0x20
reg_overflow_cnt    EQU 0x21
reg_repetition_cnt  EQU 0x22

delay_10ms_t0_overflows EQU 0x14    ; prescaler T0 set to 1:2 @ 4MHz

; -----------------------------------------------------------------------

; code memory
 org    0x0000
    clrf    STATUS      ; 00h Page 0, Bank 0
    nop                 ; 01h
    nop                 ; 02h
    goto    start       ; 03h Initialiizing / begin program

ISR org    0x0004  ; jump here on interrupt with GIE set (should not appear)
    return      ; return with GIE unset

idle  org    0x0005
    bcf   T1CON, TMR1ON
    M_skipnext_rst_notpressed
    goto  check_rst_button
    btfss GPIO, CTRL_P9
    goto  check_ctrl
    bcf   INTCON, GPIF
    sleep

    btfss GPIO, CTRL_P9 ; interupt caused by controller?
    goto  check_ctrl    ; yes -> goto controller loop

check_rst_button        ; no -> check reset button
    call  delay_10ms    ; debounce
    call  delay_10ms    ; debounce
    M_skipnext_rst_pressed
    goto  idle
    M_pause_CPU

check_rst_button_loop
    M_skipnext_rst_notpressed
    goto  check_rst_button_loop
    M_run_CPU
    goto  idle
    
check_ctrl ; this point is entered if Pin 9 of the controller port goes low
    btfsc GPIO, CTRL_P6 ; P6 also low?
    goto  idle          ; no -> back to idle
    ; P6 also low -> prepare the 1.5s (indeed, the time is around 1.573ms)
    M_movlf 0x03, reg_overflow_cnt
    M_T1reset           ; start timer 1

check_ctrl_loop
    btfsc   GPIO, CTRL_P6       ; P6 still low?
    goto    idle                ; no -> back to idle
    btfsc   GPIO, CTRL_P9       ; P9 still low?
    goto    idle                ; no -> back to idle
    btfss   PIR1, TMR1IF        ; timer 1 overflow?
    goto    check_ctrl_loop
    clrf    PIR1                ; clear overflow bit
    decfsz  reg_overflow_cnt, 1 ; Are all loops done?
    goto    check_ctrl_loop     ; If no, repeat this loop
    bcf     T1CON, TMR1ON
    
do_reset ; St+A+B+C were held for 1.5s
    M_push_reset        ; this macro pauses the CPU, too
    M_delay_x10ms 0x08
    btfss GPIO, EXTMODE ; extended mode enabled?
    goto  reset_end     ; no -> simply reset
    
ext_mode_loop
    btfsc   GPIO, CTRL_P6 ; P6 still low?
    goto    reset_end     ; no -> end_loop
    btfss   GPIO, CTRL_P9 ; P9 still low?
    goto    ext_mode_loop ; yes -> stay in this loop
    
reset_end
    M_release_reset     ; this macro starts the CPU, too
    M_delay_x10ms 0x3c  ; time (600ms) for the user to release the button comb.
    goto idle
          
delay_10ms
    clrf    TMR0                ; start timer (operation clears prescaler of T0 to 1:2)
    M_movlf delay_10ms_t0_overflows, reg_overflow_cnt
    bsf     INTCON, T0IE        ; enable timer 0 interrupt

delay_10ms_loop_pre
    bcf     INTCON, T0IF

delay_10ms_loop
    btfss   INTCON, T0IF
    goto    delay_10ms_loop
    decfsz  reg_overflow_cnt, 1
    goto    delay_10ms_loop_pre
    bcf     INTCON, T0IE        ; disable timer 0 interrupt
    return

delay_x10ms
    call    delay_10ms
    decfsz  reg_repetition_cnt, 1
    goto    delay_x10ms
    return
    
; initialisation    
start
    clrf    GPIO
    M_movlf 0x07, CMCON     ; GPIO2..0 are digital I/O (not connected to comparator)
    M_movlf 1<<GPIE, INTCON ; react on GPIE (to detect possibly pressed button combination)
    banksel TRISIO
  if Calibrate_OSCCAL
    call    3FFh            ; Get the osccal value
  else
    movlf   0xfc
  endif
    movwf   OSCCAL          ; Calibrate
    M_movlf 0x3F, TRISIO    ; in in in in in in
    M_movlf 0x09, IOC       ; IOC at GPIO3 (CTRL_P9) and GPIO0 (RST_BUTTON)
    M_movlf 0x06, WPU       ; pull-ups at GPIO2 (n.c.) and GPIO1 (CPU /HALT)
    clrf    OPTION_REG      ; global pull-ups enabled, prescaler T0 1:2
    banksel GPIO
    M_movlf 0x30, T1CON     ; set prescaler T1 1:8

detect_reset_type
    clrf    reg_reset_type
    btfss   GPIO, RST_BUTTON             ; skip next for low-active reset
    bsf     reg_reset_type, RST_BUTTON
    goto    idle

theend
    END
; ------------------------------------------------------------------------


MS_FM_BIOS/MasterSystem_FM_BIOS.asm
#include <p16f630.inc>

;
; -----------------------------------------------------------------------
;
;	Sega Master System switchless mod
;   supports:   Dual-BIOS (Eu and Jap) - needs an electrinic switch,
;                                        e.g., using a 74*125
;               FMSEB-PCB from etim - see link below
;
;   Copyright (C) 2013 by Peter Bartmann
;
;   This program is free software; you can redistribute it and/or modify
;   it under the terms of the GNU General Public License as published by
;   the Free Software Foundation; version 2 of the License only.
;
;   This program is distributed in the hope that it will be useful,
;   but WITHOUT ANY WARRANTY; without even the implied warranty of
;   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;   GNU General Public License for more details.
;
;   You should have received a copy of the GNU General Public License
;   along with this program; if not, write to the Free Software
;   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
;
; -----------------------------------------------------------------------
;
;   pin configuration: 
;
;                         ,-----_-----.
;                     +5V |1        14| GND
;      /BIOS_EU  (in/out) |2  A5 A0 13| /RESET (in) (from Reset Button)
;      /BIOS_JAP (in/out) |3  A4 A1 12| VIDEOMODE (out) (50/60Hz)
;      /BIOS_SUPP    (in) |4  A3 A2 11| /RESET (in/out) (line to Console)
;       (green) LED (out) |5  C5 C0 10| /FMSEB_EN (in/out) 
;         (red) LED (out) |6  C4 C1  9| /FMSEB_SUPP (in)
;           LED_TYPE (in) |7  C3 C2  8| /JFMSEB_EN (in/out)
;                         `-----------'
;
; Special purposes for pins:
;
;   Pin 4 (/BIOS_SUPP) sets the support for possible BIOS-pack
;      low = BIOS-pack is present - you shall connect pin 2 (/BIOS_EU) and
;                                   pin 3 (/BIOS_JAP) to pin 20 of the current
;                                   BIOS-chip
;      high = BIOS-pack is not present - pin 2 and pin 3 stay in input mode with
;										 weak pull-ups enabled
;
;   Pin 7 (LED_TYPE) sets the output mode for the LED pins
;   (must be tied to either level):
;      low  = common cathode
;      high = common anode   (output inverted)
;
;   Pin 9 (/FMSEB_SUPP) sets the support for possible BIOS-pack
;      low = FMSEB is present - you shall connect pin 8 (/JFMSEB_EN) and 
;                               pin 10 (/FMSEB_EN) to the corresponding pins of
;                               the Sega Master System FM Sound Expansion Board
;      high = FMSEB is not present - you should not connect pin 8 and pin 10
;                                    to anything else as they stay in output
;                                    mode
;
; more information about the FMSEB:
; http://members.iinet.net.au/~stinkyfist/reviletim/smsfm/smsfm.html
;
; more information about the BIOS pack:
; http://www.smspower.org/forums/viewtopic.php?t=10908
;
; -----------------------------------------------------------------------
;
; mode description:
;
; mode 1 = PAL:           50Hz, LED green,
;                         jap. BIOS not active, FMSEB not active
;
; mode 2 = NTSC:          60Hz, LED red,
;                         jap. BIOS not active, FMSEB not active
;
; mode 3 = Fake JAP NTSC: 60Hz, LED orange,
;                         jap. BIOS not active, FMSEB with FM_EN#-Signal active
;
; mode 4 = JAP NTSC:      60Hz, LED orange,
;                         jap. BIOS active, FMSEB with JPFM_EN#-Signal active
;
;
;
; NB: - All four modes are available if the FMSEB is present (pin 9 on low).
;     - The jap. BIOS can only be activated if the BIOS pack is present
;       (pin 4 on low).
;     - If mode 3 and 4 are available the LED flashes three times. If the 
;       FMSEB is not present but the BIOS pack, i.e., mode 1, 2 and 4 are
;       available, the LED does not flash.
;     - If there is a mode change, where the BIOS- or the FMSEB mode is also
;       changed, the console requires a hard reset
;       (off and on over the power button).
;       If this is needed the console shows you you have to do so by flashing
;       the LED.
;       (current color -> off -> red -> green -> orange -> off -> current color)
;       Mode changes between 50Hz and 60Hz are done while the console is
;       running.
;      
; -----------------------------------------------------------------------
; Configuration bits: adapt to your setup and needs

    __CONFIG _INTRC_OSC_NOCLKOUT & _WDT_OFF & _PWRTE_OFF & _MCLRE_OFF & _CP_OFF & _CPD_OFF

; -----------------------------------------------------------------------
; macros and definitions

M_movff macro   fromReg, toReg  ; move filereg to filereg
        movfw   fromReg
        movwf   toReg
        endm

M_movpf macro   fromPORT, toReg ; move PORTx to filereg
        movfw   fromPORT
        andlw   0x3f
        movwf   toReg
        endm

M_movlf macro   literal, toReg  ; move literal to filereg
        movlw   literal
        movwf   toReg
        endm

M_beff  macro   compReg1, compReg2, branch  ; branch if two fileregs are equal
        movfw   compReg1
        xorwf	compReg2, w
        btfsc   STATUS, Z
        goto    branch
        endm

M_bepf  macro   compPORT, compReg, branch   ; brach if PORTx equals compReg (ignoring bit 6 and 7)
        movfw   compPORT
        xorwf   compReg, w
        andlw   0x3f
        btfsc   STATUS, Z
        goto    branch
        endm

M_belf  macro   literal, compReg, branch  ; branch if a literal is stored in filereg
        movlw   literal
        xorwf	compReg, w
        btfsc   STATUS, Z
        goto    branch
        endm

M_celf  macro   literal, compReg, call_func  ; call if a literal is stored in filereg
        movlw   literal
        xorwf	compReg, w
        btfsc   STATUS, Z
        call    call_func
        endm

M_delay_x10ms   macro   literal ; delay about literal x 10ms
                movlw   literal
                movwf   reg_repetition_cnt
                call    delay_x10ms
                endm

M_push_reset    macro
                banksel TRISA
                bcf     TRISA, NRESET_OUT
                banksel PORTA
                bcf     PORTA, NRESET_OUT
                endm

M_release_reset macro
                bsf     PORTA, NRESET_OUT
                banksel TRISA
                bsf     TRISA, NRESET_OUT
                banksel PORTA
                endm

M_setBIOS_EU    macro
                banksel TRISA
                bcf     TRISA, NBIOS_EU
                bsf     TRISA, NBIOS_JAP
                banksel PORTA
                bcf     PORTA, NBIOS_EU
                endm

M_setBIOS_JAP   macro
                banksel TRISA
                bcf     TRISA, NBIOS_JAP
                bsf     TRISA, NBIOS_EU
                banksel PORTA
                bcf     PORTA, NBIOS_JAP
                endm

M_setFMSEB_EN   macro   ; led has to be set afterwards
                clrf    PORTC
                banksel TRISC
                bcf     TRISC, NFMSEB_EN
                bsf     TRISC, NJFMSEB_EN
                banksel PORTC
                endm

M_setJFMSEB_EN  macro   ; led has to be set afterwards
                clrf    PORTC
                banksel TRISC
                bcf     TRISC, NJFMSEB_EN
                bsf     TRISC, NFMSEB_EN
                banksel PORTC
                endm
                

#define M_skipnext_bios_present     btfsc   PORTA, NBIOS_SUPP
#define M_skipnext_bios_notpresent  btfss   PORTA, NBIOS_SUPP
#define M_skipnext_fm_present       btfsc   PORTC, NFMSEB_SUPP
#define M_skipnext_fm_notpresent    btfss   PORTC, NFMSEB_SUPP

#define M_set50 bsf PORTA, VIDEOMODE
#define M_set60 bcf PORTA, VIDEOMODE

#define M_skipnext_rst_pressed      btfsc   PORTA, NRESET_BUTTON
#define M_skipnext_rst_notpressed   btfss   PORTA, NRESET_BUTTON

; -----------------------------------------------------------------------

;port a
NRESET_BUTTON   EQU 0
VIDEOMODE       EQU 1
NRESET_OUT      EQU 2
NBIOS_SUPP      EQU 3
NBIOS_JAP       EQU 4
NBIOS_EU        EQU 5

;port c
NFMSEB_EN   EQU 0
NFMSEB_SUPP EQU 1
NJFMSEB_EN  EQU 2
LED_TYPE    EQU 3
LED_RED     EQU 4
LED_GREEN   EQU 5

; registers
reg_overflow_cnt    EQU 0x20
reg_repetition_cnt  EQU 0x21
reg_current_mode    EQU 0x30
reg_previous_mode   EQU 0x31
reg_led_buffer      EQU 0x40

; codes and bits
code_pal            EQU 0x00
code_ntsc           EQU 0x01
code_ntsc_fm        EQU 0x02
code_ntsc_jfm_jbios EQU 0x03

default_mode    EQU code_ntsc

bit_videomode       EQU 0     
bit_special_modes   EQU 1
bit_mode_overflow   EQU 2

code_led_off    EQU 0x00
code_led_green  EQU (1<<LED_GREEN)
code_led_red    EQU (1<<LED_RED)
code_led_yellow EQU code_led_green ^ code_led_red

code_led_invert EQU code_led_green ^ code_led_red

;mode_delay_t0_overflows EQU 0x12    ; prescaler T0 set to 1:256
delay_10ms_t0_overflows EQU 0x0a    ; prescaler T0 set to 1:4 @ 4MHz
repetitions_100ms       EQU 0x0a
repetitions_200ms       EQU 0x14
repetitions_300ms       EQU 0x1e
repetitions_mode_delay  EQU 0x4a    ; around 740ms

; -----------------------------------------------------------------------

; code memory
 org    0x0000
    clrf  STATUS  ; 00h Page 0, Bank 0
    nop           ; 01h
    nop           ; 02h
    goto  start   ; 03h begin program / Initializing

 org    0x0004  ; jump here on interrupt with GIE set (should not appear)
    return      ; return with GIE unset

 org    0x0005
idle
    M_skipnext_rst_notpressed
    goto    check_rst
    bcf     INTCON, RAIF
;    sleep

idle_wait
    btfss   INTCON, RAIF
    goto    idle_wait

check_rst
    call    delay_10ms                      ; software debounce
    call    delay_10ms                      ; software debounce
    M_skipnext_rst_pressed
    goto    idle

    M_movlf repetitions_mode_delay, reg_repetition_cnt

check_rst_loop
    call    delay_10ms
    M_skipnext_rst_pressed
    goto    doreset
    decfsz  reg_repetition_cnt, 1
    goto    check_rst_loop
    
next_mode
    incf    reg_current_mode, 1
    btfss   reg_current_mode, bit_special_modes
    goto    next_mode_end
    M_skipnext_fm_notpresent
    goto    next_mode_end
    incf    reg_current_mode, 1
    M_skipnext_bios_present
    incf    reg_current_mode, 1

next_mode_end
    btfsc   reg_current_mode, bit_mode_overflow
    clrf    reg_current_mode

    M_movlf repetitions_mode_delay, reg_repetition_cnt
    call    setled

mode_delay_loop
    call    delay_10ms
    M_skipnext_rst_pressed
    goto    apply_mode
    decfsz  reg_repetition_cnt, 1
    goto    mode_delay_loop
    goto    next_mode

doreset
    M_push_reset
    M_delay_x10ms   repetitions_300ms
    M_release_reset
    goto    idle

apply_mode
    call    save_mode
    ; strategie: mode can only be changed in non special modes
    btfsc   reg_previous_mode, bit_special_modes
    goto    check_hard_reset_needed
    btfsc   reg_current_mode, bit_special_modes
    goto    show_hard_reset_needed

    btfss   reg_current_mode, bit_videomode
    M_set50
    btfsc   reg_current_mode, bit_videomode
    M_set60
    goto    idle

check_hard_reset_needed
    M_beff  reg_current_mode, reg_previous_mode, idle   ; nothing has changed

show_hard_reset_needed
    movfw   PORTC
    andlw   0x30
    movwf   reg_led_buffer
    M_delay_x10ms   repetitions_100ms
    call    setled_off
    M_delay_x10ms   repetitions_300ms
    call    setled_green
    M_delay_x10ms   repetitions_300ms
    call    setled_red
    M_delay_x10ms   repetitions_300ms
    call    setled_yellow
    M_delay_x10ms   repetitions_300ms
    call    setled_off
    M_delay_x10ms   repetitions_300ms
    M_movff reg_led_buffer, PORTC
    goto    idle

; --------calls--------
setled
    M_belf  code_pal, reg_current_mode, setled_green
    M_belf  code_ntsc, reg_current_mode, setled_red
    M_belf  code_ntsc_fm, reg_current_mode, setled_yellow
    M_belf  code_ntsc_jfm_jbios, reg_current_mode, setled_yellow_flash

setled_off
    movlw   code_led_off
    btfsc   PORTC, LED_TYPE ; if common anode:
    xorlw   code_led_invert ; invert output
    movwf   PORTC
    return

setled_green
    movlw   code_led_green
    btfsc   PORTC, LED_TYPE ; if common anode:
    xorlw   code_led_invert ; invert output
    movwf   PORTC
    return

setled_red
    movlw   code_led_red
    btfsc   PORTC, LED_TYPE ; if common anode:
    xorlw   code_led_invert ; invert output
    movwf   PORTC
    return

setled_yellow
    movlw   code_led_yellow
    btfsc   PORTC, LED_TYPE ; if common anode:
    xorlw   code_led_invert ; invert output
    movwf   PORTC
    return

setled_yellow_flash
    M_skipnext_fm_present
    goto            setled_yellow
;    call            setled_yellow
;    M_delay_x10ms   repetitions_300ms
    call            setled_off
    M_delay_x10ms   repetitions_300ms
    call            setled_yellow
    M_delay_x10ms   repetitions_300ms
    call            setled_off
    M_delay_x10ms   repetitions_300ms
    call            setled_yellow
    M_movlf repetitions_300ms, reg_repetition_cnt   ; short the mode delay time (this line has no effect on startup)
    return

save_mode
    movfw   reg_current_mode
    banksel EEADR
    movwf   EEDAT
    bsf     EECON1,WREN
    movlw   0x55
    movwf   EECON2
    movlw   0xaa
    movwf   EECON2
    bsf     EECON1, WR
    banksel PORTA
    return


delay_10ms
    M_movlf delay_10ms_t0_overflows, reg_overflow_cnt
    clrf    W
    movwf   TMR0    ; start timer

delay_10ms_loop_pre
    bcf     INTCON, T0IF

delay_10ms_loop
    btfss   INTCON, T0IF
    goto    delay_10ms_loop
    decfsz  reg_overflow_cnt, 1
    goto    delay_10ms_loop_pre
    return

delay_x10ms
    call    delay_10ms
    decfsz  reg_repetition_cnt, 1
    goto    delay_x10ms
    return

; --------initialization--------

start
    clrf    PORTA
    clrf    PORTC
    M_movlf 0x07, CMCON                 ; GPIO2..0 are digital I/O (not connected to comparator)
    M_movlf 0x28, INTCON                ; enable interrupts: T0IE and RAIE
    banksel TRISA                       ; Bank 1
    call    3FFh                        ; Get the cal value
    movwf   OSCCAL                      ; Calibrate
    M_movlf 0x3d, TRISA                 ; in in in in out in
    M_movlf 0x0f, TRISC                 ; out out in in in in
    M_movlf (1<<NRESET_BUTTON), IOCA    ; IOC at reset button
    M_movlf 0x31, WPUA                  ; weak pullup on BIOS pins and reset button
    M_movlf 0x01, OPTION_REG            ; global pullup disable, prescaler T0 1:4
    banksel PORTA                       ; Bank 0

load_mode
    clrf    reg_current_mode
    clrf    reg_previous_mode
    bcf     STATUS, C           ; clear carry
    banksel EEADR               ; fetch current mode from EEPROM
    clrf    EEADR               ; address 0
    bsf     EECON1, RD
    movf    EEDAT, w
    banksel PORTA
    movwf   reg_current_mode    ; last mode saved
    movwf   reg_previous_mode   ; last mode saved to compare

set_initial_mode
    ; strategie: try to set mode, set if valid, set to default if not
    M_belf  code_pal, reg_current_mode, set_pal
    M_belf  code_ntsc, reg_current_mode, set_ntsc
    M_belf  code_ntsc_fm, reg_current_mode, set_ntsc_fm
    M_belf  code_ntsc_jfm_jbios, reg_current_mode, set_ntsc_jap
    goto    set_default  ; should not appear

set_pal
    M_set50
    M_setBIOS_EU
    goto    init_end

set_default
    movlw   default_mode
    movwf   reg_current_mode
    movwf   reg_previous_mode

set_ntsc
    M_set60
    M_setBIOS_EU
    goto    init_end

set_ntsc_fm
    M_skipnext_fm_present
    goto    set_default
    M_setFMSEB_EN
    M_setBIOS_EU
    goto    init_end

set_ntsc_jap
    M_skipnext_fm_present
    goto    set_ntsc_jbios_only
    M_setJFMSEB_EN
    M_setBIOS_JAP
    goto    init_end

set_ntsc_jbios_only
    M_skipnext_bios_present
    goto    set_default
    M_setBIOS_JAP

init_end
    call    save_mode
    call    setled
    goto    idle

; -----------------------------------------------------------------------
; eeprom data
DEEPROM	CODE
	de	default_mode

theend
    END
; ------------------------------------------------------------------------
